# Document Similarity

In [1]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity


documents = ['Document one text here', 'Document two text goes here']

tfidf_vectorizer = TfidfVectorizer()
tfidf_matrix = tfidf_vectorizer.fit_transform(documents)

similarity = cosine_similarity(tfidf_matrix[0:1], tfidf_matrix)
print(similarity)

[[1.         0.51014902]]


The output [[1. 0.51014902]] from the cosine similarity function represents a similarity matrix where 1 indicates perfect similarity of a document with itself, and 0.51014902 shows moderate similarity between two different documents. The closer the value is to 1, the more similar the documents are.

In [2]:
documents = [
    "machine learning is powerful",
    "AI and machine learning are useful",
    "football is a popular sport"
]

# Convert documents to TF-IDF vectors
vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(documents)

# Compute cosine similarity
similarity_matrix = cosine_similarity(tfidf_matrix)

print(similarity_matrix)

[[1.         0.30801606 0.18487962]
 [0.30801606 1.         0.        ]
 [0.18487962 0.         1.        ]]


### Using Gensim Word2Vec

In [14]:
from gensim.models import Word2Vec
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity

docs = [
    "machine learning is powerful",
    "AI and machine learning are powerful"
]

#tokenize
tokenized_docs = [doc.split() for doc in docs]
model = Word2Vec(sentences=tokenized_docs, vector_size=100, window=5, min_count=1)

def document_vector(doc):
    vectors = [model.wv[word] for word in doc if word in model.wv]
    return np.mean(vectors, axis=0)

doc1_vec = document_vector(tokenized_docs[0])
doc2_vec = document_vector(tokenized_docs[1])

similarity = cosine_similarity([doc1_vec], [doc2_vec])

print("similarity:", similarity[0][0])

similarity: 0.5988256


### Using Sentence embeddings instead of Word

Instead of averaging words, models directly produce document vectors.

Example models:

Sentence-BERT

Universal Sentence Encoder

In [17]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

model = SentenceTransformer("all-MiniLM-L6-v2")

docs = [
    "machine learning is powerful",
    "AI and machine learning are useful"
]

embeddings = model.encode(docs)

similarity = cosine_similarity([embeddings[0]], [embeddings[1]])

print(similarity)

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/all-MiniLM-L6-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

[[0.6942031]]


# Document Search Engine

In [ ]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

documents = [
    "Machine learning is used in AI.",
    "Deep Learning improves computer vision",
    "Football is a popular sport and Messi is a Greates of all time in it.",
    "Natural Language processing uses machine Learning"
]

In [21]:
vectorizer = TfidfVectorizer()
doc_vectors = vectorizer.fit_transform(documents)

In [22]:
def search(query):
    
    #convert query to vector
    query_vec = vectorizer.transform([query])

    #compute similarity
    similarity = cosine_similarity(query_vec, doc_vectors)

    scores = similarity.flatten()
    ranked_docs = scores.argsort()[::-1]


    print("Seacrch Results:\n")
    for i in ranked_docs[:3]:
        print(documents[i], "| score:", scores[i])

In [32]:
search("ai")

Seacrch Results:

Machine learning is used in AI. | score: 0.4595480329387006
Football is a popular sport and Messi is a Greates of all time in it. | score: 0.0
Deep Learning improves computer vision | score: 0.0
